src/main/java/com/google/api/codegen/Resources.java

```diff
@@ -69,22 +72,48 @@
     // (useful for testability).
     Map<String, FieldSegment> specs = new TreeMap<>();
 
+    Map<String, List<FieldSegment>> methodFieldSegs = getMethodFieldSegmentMapFromHttpPaths(methods);
+    for (List<FieldSegment> segList : methodFieldSegs.values()) {
+      for (FieldSegment fieldSegment : segList) {
+        String resourcePath = PathSegment.toSyntax(fieldSegment.getSubPath());
+        // If there are multiple field segments with the same resource path, the last
+        // one will be used, making the output deterministic. Also, the first field path
+        // encountered tends to be simply "name" because it is the corresponding create
+        // API method for the type.
+        specs.put(resourcePath, fieldSegment);
+      }
+    }
+    return specs.values();
+  }
+
+  /**
+   * Returns a map from method name to a list of FieldSegment objects. Each FieldSegment will be
+   * composed of a field name and a path segment, representing an abstract resource name with
+   * wildcards.
+   */
+  public static Map<String, List<FieldSegment>> getMethodFieldSegmentMapFromHttpPaths(
+      List<Method> methods) {
+    Map<String, List<FieldSegment>> fieldSegmentsPerMethodMap =
+        new Hashtable<String, List<FieldSegment>>();
+
     for (Method method : methods) {
+      fieldSegmentsPerMethodMap.put(method.getFullName(), new ArrayList<FieldSegment>());
       HttpAttribute httpAttr = method.getAttribute(HttpAttribute.KEY);
       for (PathSegment pathSegment : httpAttr.getPath()) {
         if (isTemplateFieldSegment(pathSegment)) {
           FieldSegment fieldSegment = (FieldSegment) pathSegment;
-          String resourcePath = PathSegment.toSyntax(fieldSegment.getSubPath());
-          // If there are multiple field segments with the same resource path, the last
-          // one will be used, making the output deterministic. Also, the first field path
-          // encountered tends to be simply "name" because it is the corresponding create
-          // API method for the type.
-          specs.put(resourcePath, fieldSegment);
+          fieldSegmentsPerMethodMap.get(method.getFullName()).add(fieldSegment);
         }
       }
     }
 
-    return specs.values();
+    return fieldSegmentsPerMethodMap;
+  }
+
+  public static String getEntityName(FieldSegment segment) {
+    // TODO(shinfan): Consider finding a better way to determine the name if possible.
+    List<String> params = Lists.newArrayList(Resources.getParamsForResourceNameWildcards(segment));
```

Comment:
getParamsForResourceNameWildcards is only being used here. I think it would simplify things to just merge the methods - this getEntityName has to contort itself a little bit because of the return type of getParamsForResourceNameWildcards, which is silly.